### PR TITLE
datapath: Optionally accelerate NodePort translation

### PR DESCRIPTION
**Pull request by brb**
The commit 4ac28ce2b ("bpf: make bpf_sock REMOTE_NODE_ID aware") introduced the following optimization in NodePort service translation:

    If a client which runs on any node managed by Cilium sends a request to
    a NodePort service via another node (i.e. dst IP != the client node
    IP), then the NodePort service lookup and translation will happen on
    the client-side instead of the destination node.

Unfortunately, the optimization complicates some test cases of `test/k8sT/Services.go` suite, in which we expect that the lookup and translation will happen on the destination node (we want to test `bpf_netdev.c` instead of `bpf_sock.c`).

To simplify the testing, introduce `--enable-node-port-xlr flag` (defaults to `true`) which optionally enables the optimization.

Happy to consider a better flag name.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9877)

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -172,8 +172,8 @@ static inline bool sock4_is_external_ip(struct lb4_service *svc,
 
 		info = ipcache_lookup4(&IPCACHE_MAP, key->address,
 				       V4_CACHE_KEY_LEN);
-		if (info == NULL || (info->sec_label != HOST_ID &&
-				     info->sec_label != REMOTE_NODE_ID))
+		if (info == NULL || (info->sec_label != HOST_ID
+				     && info->sec_label != REMOTE_NODE_ID))
```
Comment (borkmann) [inline bpf/bpf_sock.c]: ^ above diff slipped in?

**Comment by brb** [comment]
test-me-please

File: bpf/bpf_sock.c [hunk 2]
```diff
@@ -201,9 +201,14 @@ static inline void sock4_handle_node_port(struct bpf_sock_addr *ctx,
...
 
 	info = ipcache_lookup4(&IPCACHE_MAP, daddr, V4_CACHE_KEY_LEN);
-	if (info != NULL && (info->sec_label == HOST_ID ||
-			     info->sec_label == REMOTE_NODE_ID))
+	if (info != NULL && (info->sec_label == HOST_ID
+#ifdef ENABLE_NODEPORT_XLR
+			     || info->sec_label == REMOTE_NODE_ID)) {
+#else
+	)) {
+#endif
```
Comment (borkmann) [inline bpf/bpf_sock.c:209]: Lets make this a proper helper function e.g. `sock_can_xlate_direct(info->sec_label)` and move the ifdef in there. This should be applied to NodePort and ExternalIP.

Comment (borkmann) [reply bpf/bpf_sock.c:209]: Commit desc says: `To simplify the testing, introduce --enable-node-port-xlr flag (defaults
to "true") which optionally enables the optimization.` -> `optionally disables`

Comment (borkmann) [reply bpf/bpf_sock.c:209]: Can we find a better name for `--enable-node-port-xlr` as this should cover NodePort and ExternalIP?

**Comment by brb** [comment]
Had a second thought on this - changing the core datapath to make testing easier feels a bit wrong. I will fix the CI provisioning, and meanwhile I will disable the optimization to avoid breaking NP in bpf_netdev.c until the provisioning of the third host is fixed (in a separate PR).

**Pull request closed without merging**